### PR TITLE
DEPR: pandas.util.testing

### LINKED ISSUE #16232: DEPR: pandas.core, pandas.util.testing
xref #16223 

These are the only remaining 'public'-ish directories that are not explicitly exposed. In 0.20.0 we make these 'private' in documentation. We should actually deprecate these and move to ``_core`` and ``pandas.util.testing`` -> ``pandas.core.testing``

This will need a DeprecationWarning as likely lots of code that is out there that uses this.

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/16232.

I tried to keep the commits somewhat clean.

https://github.com/pandas-dev/pandas/commit/e96624b8a3f903798f7977fb92da09fc7417ec98 and https://github.com/pandas-dev/pandas/commit/d52d35f9d6a5e721ae0e6aa1579e44dac820ed7a can probably be ignored. That's just moving `pandas.util.testing` to `pandas.util._testing` and updating all the relevant imports.

Main question: do we want to make the `tm.make*` methods public? I'd say we probably should, but wanted to confirm.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
On second thought, we may want to hold of on making the `make*` functions public. We'll want pep8 compatible names at least, possibly other changes.

**Comment by jbrockmendel** [comment]
The naming is problematic for our isort config.  We treat `pandas.util._*` as a collection of low-dependency modules, which `tm` is decidedly not.  could go directly in pandas/_testing.py?

**Comment by TomAugspurger** [comment]
Moved to pandas._testing.

**Comment by TomAugspurger** [comment]
Not sure about the mypy failure.

```
mypy --version
mypy 0.730
Performing static analysis using mypy
pandas/tests/test_register_accessor.py:39: error: Module has no attribute "register_series_accessor"
pandas/tests/test_register_accessor.py:40: error: Module has no attribute "register_dataframe_accessor"
pandas/tests/test_register_accessor.py:41: error: Module has no attribute "register_index_accessor"
Found 3 errors in 1 file (checked 839 source files)
```

I saw this in https://github.com/pandas-dev/pandas/pull/29964 too, but have never reproduced locally.

**Comment by jbrockmendel** [comment]
I see it locally occasionally.  Seems pretty random.

**Comment by jreback** [comment]
lgtm.

**Review by jreback — APPROVED** [review]
+1 merge on green.

**Comment by WillAyd** [comment]
Haven't seen the error before - does merging master resolve?

> Main question: do we want to make the `tm.make*` methods public? I'd say we probably should, but wanted to confirm.

I don't think we want this. I'm under the impression these are better converted to fixtures

**Comment by TomAugspurger** [comment]
Sounds good. Planning to merge this once it passes since it touch so many files.

**Comment by TomAugspurger** [comment]
Another mypy failure.

**Comment by simonjayhawkins** [comment]
> Not sure about the mypy failure.
> 
> ```
> mypy --version
> mypy 0.730
> Performing static analysis using mypy
> pandas/tests/test_register_accessor.py:39: error: Module has no attribute "register_series_accessor"
> pandas/tests/test_register_accessor.py:40: error: Module has no attribute "register_dataframe_accessor"
> pandas/tests/test_register_accessor.py:41: error: Module has no attribute "register_index_accessor"
> Found 3 errors in 1 file (checked 839 source files)
> ```
> 
> I saw this in #29964 too, but have never reproduced locally.

i see this locally quite often on my check_untyped_defs branch. can normally clear with `--no-incremental` but that's not relevant for ci since there isn't a cache.

**Comment by TomAugspurger** [comment]
Not sure if it's relevant, but we seem to only import `pandas.api` via `pandas.io.json._table_schema`, which doesn't seem great. I'm going to restructure things a bit in another PR.

**Comment by TomAugspurger** [comment]
Passing now. Merging.

File: doc/source/reference/general_utility_functions.rst
Comment (jorisvandenbossche) [inline doc/source/reference/general_utility_functions.rst]: Is it needed to make all those asserts public? 
Eg, what's the pandas-specific value for external projects of using pandas' `assert_equal` or `assert_almost_equal` ?
Also all the EA specific ones could probably be limited to a single general array assert?

Comment (jorisvandenbossche) [reply doc/source/reference/general_utility_functions.rst]: Thoughts about this? (can also make an issue for it)

**Pull request merged**
